File: matterbase/query.py
"""The constructed query — matterbase's actual purpose. No Textual dependency.

QueryState holds the three channels of the query builder:

* grubber presets (the "sets") — become ``-f`` expressions,
* SQL WHERE — including filename search folded in as SQL (``_note_file LIKE``),
* full-text — a *display filter* with special status: it narrows what is shown
  but is never part of the yankable command (grubber | duckdb cannot express it).
"""

import re
import shlex
from dataclasses import dataclass, field
from pathlib import Path

from .content import source_type


@dataclass
class Preset:
    """A named grubber query ("set") from the config."""
    label: str
    exprs: list[str]
    active: bool = False


@dataclass
class QueryState:
    presets: list[Preset] = field(default_factory=list)
    sql_where: str = ""
    filename_term: str = ""
    fulltext_term: str = ""

    # ── grubber channel ──────────────────────────────────────────────

    def active_expressions(self) -> list[str]:
        """All active presets' expressions, flattened into one grubber call
        (record-level AND) — identical semantics in display and yank."""
        return [e for p in self.presets if p.active for e in p.exprs]

    # ── SQL channel (filename folded in) ─────────────────────────────

    def filename_clause(self) -> str:
        """The filename search expressed as SQL, or "" when inactive."""
        term = self.filename_term.strip()
        if not term:
            return ""
        escaped = term.replace("'", "''")
        return f"_note_file LIKE '%{escaped}%'"

    def effective_sql(self) -> str:
        """User SQL AND the filename clause — the WHERE the pipeline runs."""
        user = self.sql_where.strip()
        fname = self.filename_clause()
        if user and fname:
            return f"({user}) AND {fname}"
        return user or fname

    # ── full-text channel (display only, never yanked) ───────────────

    def fulltext_active(self) -> bool:
        return bool(self.fulltext_term.strip())

    # ── the constructed command ──────────────────────────────────────

    def build_command(
        self,
        notes_dir: str,
        *,
        search_mode: str = "all",
        mmd: bool = False,
        depth: int | None = None,
        collection_dir: str | None = None,
        array_fields: list[str] | None = None,
        grubber_set: str = "",
    ) -> str:
        """The yankable `grubber … | duckdb 'SQL'` pipeline.

        Reproduces the *structured* query only — full-text is deliberately
        absent (decision 1: yank ≠ displayed set when full-text is active).

        With *grubber_set*, the database definition (path, JSONL sources,
        merge keys) is assumed to live in the grubber config set of that
        name, and the command shrinks to `grubber extract --set NAME …`.
        The set must describe the same database as this session's config.
        """
        if grubber_set:
            parts = [
                shlex.quote(get_grubber_bin_name()), "extract",
                "--set", shlex.quote(grubber_set),
            ]
        else:
            parts = [shlex.quote(get_grubber_bin_name()), "extract", shlex.quote(notes_dir)]
        if search_mode == "frontmatter":
            parts.append("--frontmatter-only")
        elif search_mode == "blocks_only":
            parts.append("--blocks-only")
        else:
            parts.append("-a")
        if depth is not None:
            parts += ["--depth", str(depth)]
        if mmd:
            parts.append("--mmd")
        if collection_dir and not grubber_set:
            from .grubber_client import COLLECTION_MERGE_KEYS
            parts += ["--from-jsonl", shlex.quote(collection_dir)]
            parts += ["--merge-on", COLLECTION_MERGE_KEYS]
        for expr in self.active_expressions():
            parts += ["-f", shlex.quote(expr)]
        cmd = " ".join(parts)
        if array_fields:
            cmd = f"GRUBBER_ARRAY_FIELDS={','.join(array_fields)} {cmd}"
        sql = self.effective_sql()
        if sql:
            escaped_sql = sql.replace('"', '\\"')
            cmd += (
                f' | duckdb -json -c "SELECT * FROM'
                f" read_json_auto('/dev/stdin') WHERE {escaped_sql}\""
            )
        return cmd


def get_grubber_bin_name() -> str:
    # Late import so tests can monkeypatch grubber_client.GRUBBER_BIN.
    from . import grubber_client
    return grubber_client.GRUBBER_BIN


# ---------------------------------------------------------------------------
# SQL form helper
#
# The form (field → operator → value) is comfort only: it *generates* a WHERE
# clause into the SQL input, which stays the single source of truth and is
# freely editable afterwards.
# ---------------------------------------------------------------------------

SQL_FORM_OPERATORS = (
    "=", "!=", "LIKE", ">", "<", ">=", "<=", "IS NULL", "IS NOT NULL", "IN",
)

_SIMPLE_IDENT = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*$")


def _sql_ident(field: str) -> str:
    """Quote a field name unless it is a plain SQL identifier."""
    if _SIMPLE_IDENT.match(field):
        return field
    return '"' + field.replace('"', '""') + '"'


def _sql_value(value: str) -> str:
    """Render a scalar: numbers stay bare, everything else single-quoted."""
    v = value.strip()
    try:
        float(v)
        return v
    except ValueError:
        return "'" + v.replace("'", "''") + "'"


def build_clause(field: str, op: str, value: str = "") -> str:
    """Build one WHERE clause from the form inputs. Returns "" when the
    combination is incomplete (no field, or a missing required value)."""
    field = field.strip()
    op = op.strip().upper()
    if not field or op not in SQL_FORM_OPERATORS:
        return ""
    ident = _sql_ident(field)

    if op in ("IS NULL", "IS NOT NULL"):
        return f"{ident} {op}"

    value = value.strip()
    if not value:
        return ""

    if op == "LIKE":
        # Substring search by default; explicit wildcards are kept as-is.
        pattern = value if "%" in value else f"%{value}%"
        return f"{ident} LIKE '" + pattern.replace("'", "''") + "'"

    if op == "IN":
        items = [v.strip() for v in value.split(",") if v.strip()]
        if not items:
            return ""
        return f"{ident} IN (" + ", ".join(_sql_value(v) for v in items) + ")"

    return f"{ident} {op} {_sql_value(value)}"


def append_clause(sql: str, clause: str) -> str:
    """AND a generated clause onto existing SQL (or start with it)."""
    sql = sql.strip()
    if not clause:
        return sql
    return f"{sql} AND {clause}" if sql else clause


def remove_last_clause(sql: str) -> str:
    """Strip the last top-level AND-clause; a single clause clears to "".

    Quote- and paren-aware so an ``AND`` inside a string literal or inside
    ``IN (…)`` does not count as a split point.
    """
    s = sql.strip()
    in_str = False
    depth = 0
    last = -1
    i = 0
    while i < len(s):
        c = s[i]
        if in_str:
            if c == "'":
                if i + 1 < len(s) and s[i + 1] == "'":
                    i += 1  # escaped '' stays inside the literal
                else:
                    in_str = False
        elif c == "'":
            in_str = True
        elif c == "(":
            depth += 1
        elif c == ")":
            depth -= 1
        elif depth == 0 and s[i : i + 5].upper() == " AND ":
            last = i
        i += 1
    if last < 0:
        return ""
    return s[:last].rstrip()


# ---------------------------------------------------------------------------
# Full-text display filter
#
# Searches prose + YAML-block body — what grubber | duckdb cannot express.
# Only markdown/typst records have a body; jsonl records drop out while
# full-text is active (decision 2). File contents are cached per refresh.
# ---------------------------------------------------------------------------

def filter_records_fulltext(
    records: list[dict],
    term: str,
    file_cache: dict[str, str] | None = None,
) -> list[dict]:
    """Keep only markdown/typst records whose source file contains *term*.

    file_cache maps path → lowercased content; populated lazily so each
    source file is read at most once per cache lifetime.
    """
    needle = term.strip().lower()
    if not needle:
        return records
    cache = file_cache if file_cache is not None else {}

    result: list[dict] = []
    for rec in records:
        path = rec.get("_note_file", "")
        if source_type(path) not in ("markdown", "typst"):
            continue
        content = cache.get(path)
        if content is None:
            try:
                content = Path(path).read_text(
                    encoding="utf-8", errors="replace"
                ).lower()
            except OSError:
                content = ""
            cache[path] = content
        if needle in content:
            result.append(rec)
    return result

File: cmd/matterbase/query.go
// The constructed query — port of matterbase's query.py. QueryState holds
// the three channels: grubber presets ("-f" expressions), SQL WHERE
// (filename search folded in), and full-text (display only, never
// yanked — grubber | duckdb cannot express it).
package main

import (
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
	"unicode"

	"github.com/rhsev/matterbase/basekit/recordtable"
)

// Record is one schemaless record, shared with recordtable.
type Record = recordtable.Record

// Preset is a named grubber query ("set") from the config.
type Preset struct {
	Label  string
	Exprs  []string
	Active bool
}

// QueryState is the query builder's full state.
type QueryState struct {
	Presets      []Preset
	SQLWhere     string
	FilenameTerm string
	FulltextTerm string
}

// ActiveExpressions flattens every active preset's expressions into one
// grubber call (record-level AND) — identical semantics in display and
// yank.
func (q *QueryState) ActiveExpressions() []string {
	var exprs []string
	for _, p := range q.Presets {
		if p.Active {
			exprs = append(exprs, p.Exprs...)
		}
	}
	return exprs
}

// FilenameClause is the filename search expressed as SQL, or "" when
// inactive.
func (q *QueryState) FilenameClause() string {
	term := strings.TrimSpace(q.FilenameTerm)
	if term == "" {
		return ""
	}
	return "_note_file LIKE '%" + strings.ReplaceAll(term, "'", "''") + "%'"
}

// EffectiveSQL is the user SQL AND the filename clause — the WHERE the
// pipeline actually runs.
func (q *QueryState) EffectiveSQL() string {
	user := strings.TrimSpace(q.SQLWhere)
	fname := q.FilenameClause()
	switch {
	case user != "" && fname != "":
		return "(" + user + ") AND " + fname
	case user != "":
		return user
	default:
		return fname
	}
}

// FulltextActive reports whether the full-text display filter is set.
func (q *QueryState) FulltextActive() bool {
	return strings.TrimSpace(q.FulltextTerm) != ""
}

// BuildCommandOpts are the session settings BuildCommand needs — the
// config-derived values app.py keeps on self.
type BuildCommandOpts struct {
	SearchMode    string
	MMD           bool
	Depth         *int
	CollectionDir string
	ArrayFields   []string
	GrubberSet    string
}

// BuildCommand renders the yankable `grubber … | duckdb 'SQL'` pipeline.
// Full-text is deliberately absent (decision 1: yank ≠ displayed set
// when full-text is active).
func (q *QueryState) BuildCommand(notesDir string, opts BuildCommandOpts) string {
	var parts []string
	if opts.GrubberSet != "" {
		parts = append(parts, shellQuote(grubberBin()), "extract", "--set", shellQuote(opts.GrubberSet))
	} else {
		parts = append(parts, shellQuote(grubberBin()), "extract", shellQuote(notesDir))
	}
	switch opts.SearchMode {
	case "frontmatter":
		parts = append(parts, "--frontmatter-only")
	case "blocks_only":
		parts = append(parts, "--blocks-only")
	default:
		parts = append(parts, "-a")
	}
	if opts.Depth != nil {
		parts = append(parts, "--depth", strconv.Itoa(*opts.Depth))
	}
	if opts.MMD {
		parts = append(parts, "--mmd")
	}
	if opts.CollectionDir != "" && opts.GrubberSet == "" {
		parts = append(parts, "--from-jsonl", shellQuote(opts.CollectionDir))
		parts = append(parts, "--explode", collectionExplodeField)
		parts = append(parts, "--merge-on", collectionMergeKeys)
	}
	for _, expr := range q.ActiveExpressions() {
		parts = append(parts, "-f", shellQuote(expr))
	}
	cmd := strings.Join(parts, " ")
	if len(opts.ArrayFields) > 0 {
		cmd = "GRUBBER_ARRAY_FIELDS=" + strings.Join(opts.ArrayFields, ",") + " " + cmd
	}
	if sql := q.EffectiveSQL(); sql != "" {
		escaped := strings.ReplaceAll(sql, `"`, `\"`)
		cmd += fmt.Sprintf(` | duckdb -json -c "SELECT * FROM read_json_auto('/dev/stdin') WHERE %s"`, escaped)
	}
	return cmd
}

// shellQuote is Go's shlex.quote: return s unchanged if it's already
// shell-safe, else single-quote it (escaping embedded quotes).
func shellQuote(s string) string {
	if s == "" {
		return "''"
	}
	if shellSafeRe.MatchString(s) {
		return s
	}
	return "'" + strings.ReplaceAll(s, "'", `'"'"'`) + "'"
}

var shellSafeRe = regexp.MustCompile(`^[A-Za-z0-9_@%+=:,./-]+$`)

// ---------------------------------------------------------------------------
// SQL form: field → operator → value ⇒ clause. Comfort only — it
// generates into the SQL input, which stays the single source of truth.
// ---------------------------------------------------------------------------

// sqlFormOperatorOrder is the order the SQL form's operator cycler
// steps through — matches the Python original's SQL_FORM_OPERATORS tuple.
var sqlFormOperatorOrder = []string{
	"=", "!=", "LIKE", ">", "<", ">=", "<=", "IS NULL", "IS NOT NULL", "IN",
}

var sqlFormOperators = func() map[string]bool {
	m := make(map[string]bool, len(sqlFormOperatorOrder))
	for _, op := range sqlFormOperatorOrder {
		m[op] = true
	}
	return m
}()

var simpleIdentRe = regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9_]*$`)

func sqlIdent(field string) string {
	if simpleIdentRe.MatchString(field) {
		return field
	}
	return `"` + strings.ReplaceAll(field, `"`, `""`) + `"`
}

// sqlValue renders a scalar: numbers stay bare, everything else is
// single-quoted.
func sqlValue(value string) string {
	v := strings.TrimSpace(value)
	if _, err := strconv.ParseFloat(v, 64); err == nil {
		return v
	}
	return "'" + strings.ReplaceAll(v, "'", "''") + "'"
}

// buildClause builds one WHERE clause from the SQL form inputs. Returns
// "" when the combination is incomplete.
func buildClause(field, op, value string) string {
	field = strings.TrimSpace(field)
	op = strings.ToUpper(strings.TrimSpace(op))
	if field == "" || !sqlFormOperators[op] {
		return ""
	}
	ident := sqlIdent(field)

	if op == "IS NULL" || op == "IS NOT NULL" {
		return ident + " " + op
	}

	value = strings.TrimSpace(value)
	if value == "" {
		return ""
	}

	if op == "LIKE" {
		pattern := value
		if !strings.Contains(value, "%") {
			pattern = "%" + value + "%"
		}
		return ident + " LIKE '" + strings.ReplaceAll(pattern, "'", "''") + "'"
	}

	if op == "IN" {
		var items []string
		for _, part := range strings.Split(value, ",") {
			if p := strings.TrimSpace(part); p != "" {
				items = append(items, p)
			}
		}
		if len(items) == 0 {
			return ""
		}
		vals := make([]string, len(items))
		for i, it := range items {
			vals[i] = sqlValue(it)
		}
		return ident + " IN (" + strings.Join(vals, ", ") + ")"
	}

	return ident + " " + op + " " + sqlValue(value)
}

// appendClause ANDs a generated clause onto existing SQL (or starts with
// it).
func appendClause(sql, clause string) string {
	sql = strings.TrimSpace(sql)
	if clause == "" {
		return sql
	}
	if sql != "" {
		return sql + " AND " + clause
	}
	return clause
}

// removeLastClause strips the last top-level AND-clause; a single clause
// clears to "". Quote- and paren-aware so an AND inside a string literal
// or an IN (…) list doesn't count as a split point.
func removeLastClause(sql string) string {
	s := strings.TrimSpace(sql)
	inStr := false
	depth := 0
	last := -1
	for i := 0; i < len(s); i++ {
		c := s[i]
		switch {
		case inStr:
			if c == '\'' {
				if i+1 < len(s) && s[i+1] == '\'' {
					i++
				} else {
					inStr = false
				}
			}
		case c == '\'':
			inStr = true
		case c == '(':
			depth++
		case c == ')':
			depth--
		case depth == 0 && i+5 <= len(s) && strings.EqualFold(s[i:i+5], " and "):
			last = i
		}
	}
	if last < 0 {
		return ""
	}
	return strings.TrimRightFunc(s[:last], unicode.IsSpace)
}

// ---------------------------------------------------------------------------
// Full-text display filter: searches prose + YAML-block body — what
// grubber | duckdb cannot express. Only markdown/typst records have a
// body; jsonl records drop out while full-text is active.
// ---------------------------------------------------------------------------

func filterRecordsFulltext(records []Record, term string, fileCache map[string]string) []Record {
	needle := strings.ToLower(strings.TrimSpace(term))
	if needle == "" {
		return records
	}
	if fileCache == nil {
		fileCache = map[string]string{}
	}
	var result []Record
	for _, rec := range records {
		path, _ := rec["_note_file"].(string)
		st := sourceType(path)
		if st != "markdown" && st != "typst" {
			continue
		}
		content, ok := fileCache[path]
		if !ok {
			if data, err := os.ReadFile(path); err == nil {
				content = strings.ToLower(string(data))
			}
			fileCache[path] = content
		}
		if strings.Contains(content, needle) {
			result = append(result, rec)
		}
	}
	return result
}
